resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/kubernetes/KubernetesClusterSchedulerBackend.scala

```diff
@@ -313,6 +377,101 @@ private[spark] class KubernetesClusterSchedulerBackend(
     }
   }
+
+  private class KubernetesDriverEndpoint(
+    rpcEnv: RpcEnv,
+    sparkProperties: Seq[(String, String)])
+    extends DriverEndpoint(rpcEnv, sparkProperties) {
+    override def receiveAndReply(
+      context: RpcCallContext): PartialFunction[Any, Unit] = {
+      new PartialFunction[Any, Unit]() {
+        override def isDefinedAt(msg: Any): Boolean = {
+          msg match {
+            case RetrieveSparkAppConfig(executorId) =>
+              Utils.isDynamicAllocationEnabled(sc.conf)
+            case _ => false
+          }
+        }
+
+        override def apply(msg: Any): Unit = {
+          msg match {
+            case RetrieveSparkAppConfig(executorId) =>
+              RUNNING_EXECUTOR_PODS_LOCK.synchronized {
+                var resolvedProperties = sparkProperties
+                val runningExecutorPod = kubernetesClient
+                  .pods()
+                  .withName(
+                    runningExecutorPods(executorId).getMetadata.getName)
+                  .get()
+                val nodeName = runningExecutorPod.getSpec.getNodeName
+                val shufflePodIp = shufflePodCache.get.getShufflePodForExecutor(nodeName)
+                resolvedProperties = resolvedProperties ++ Seq(
+                  (SPARK_SHUFFLE_SERVICE_HOST.key, shufflePodIp))
+
+                val reply = SparkAppConfig(
+                  resolvedProperties,
+                  SparkEnv.get.securityManager.getIOEncryptionKey())
+                context.reply(reply)
+              }
+          }
+        }
+      }.orElse(super.receiveAndReply(context))
+    }
+  }
+
+  /*
+   * KubernetesAllocator class watches executor registrations, limits
+   * and creates new executors when it is appropriate.
+   */
+  private[spark] class KubernetesAllocator(client: KubernetesClient)
+    extends Logging {
+
+    private val interval = conf.get(KUBERNETES_ALLOCATION_BATCH_DELAY)
+    private val scheduler: ScheduledExecutorService = ThreadUtils
+      .newDaemonSingleThreadScheduledExecutor("kubernetes-pod-allocator")
+
+    private val allocatorRunnable: Runnable = new Runnable {
+      override def run(): Unit = {
+        if (runningExecutorPods.size - totalRegisteredExecutors.get() > 0) {
+          logDebug("Waiting for pending executors before scaling")
+          return
+        }
+
+        if (totalExpectedExecutors.get() <= runningExecutorPods.size) {
+          logDebug(
+            "Maximum allowed executor limit reached. Not scaling up further.")
+          return
+        }
+
+        RUNNING_EXECUTOR_PODS_LOCK.synchronized {
+          for (i <- 1 to math.min(
+            totalExpectedExecutors.get - runningExecutorPods.size,
+            5)) {
+            runningExecutorPods += allocateNewExecutorPod()
+            logInfo(
+              s"Requesting a new executor, total executors is now ${runningExecutorPods.size}")
+          }
+        }
+      }
+    }
+
+    def start(): Unit = {
+      if (interval > 0) {
```

Comment:
Just make the `interval` an Optional configuration and use `interval.foreach`, as opposed to checking if an integer is greater than zero. We should validate then that any given value is positive and throw an exception if it isn't.